transformer_engine/pytorch/graph.py

```diff
@@ -350,6 +435,21 @@ def hook_fn(module, inputs, outputs):  # pylint: disable=unused-argument
 
                     per_callable_static_grad_outputs[per_callable_bwd_idx] = static_grad_outputs
                     per_callable_static_grad_inputs[per_callable_bwd_idx] = static_grad_inputs
+
+                    if io_memory_reduction:
+                        # Weak ref the static outputs and static grad inputs.
+                        per_callable_static_outputs[per_callable_bwd_idx] = make_weak_ref(
+                            static_outputs
+                        )
+                        if previous_per_callable_bwd_idx is not None:
+                            per_callable_static_grad_inputs[previous_per_callable_bwd_idx] = (
+                                make_weak_ref(
+                                    per_callable_static_grad_inputs[previous_per_callable_bwd_idx]
+                                )
+                            )
+                        previous_per_callable_bwd_idx = per_callable_bwd_idx
+                        del static_outputs, static_grad_inputs, grad_inputs
```

Comment:
We should expand this comment. The code is not intuitive enough to easily understand.

For future reference: Once a model chunk microbatch has finished its backward, we assume its output tensor is no longer needed. We can deallocate and let PyTorch's memory allocator reuse it elsewhere.